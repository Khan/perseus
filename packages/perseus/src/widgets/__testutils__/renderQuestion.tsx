import {RenderStateRoot} from "@khanacademy/wonder-blocks-core";
import {render} from "@testing-library/react";
import * as React from "react";

import {
    testDependenciesV2,
    testDependencies,
    // eslint-disable-next-line import/no-relative-packages
} from "../../../../../testing/test-dependencies";
import {
    DependenciesContext,
    useDependencies,
    setDependencies,
} from "../../dependencies";
import * as Perseus from "../../index";
import {mockStrings} from "../../strings";
import {registerAllWidgetsForTesting} from "../../util/register-all-widgets-for-testing";

import type {APIOptions} from "../../types";
import type {PerseusRenderer} from "@khanacademy/perseus-core";
import type {PropsFor} from "@khanacademy/wonder-blocks-core";

type RenderResult = ReturnType<typeof render>;

type ExtraProps = Omit<PropsFor<typeof Perseus.Renderer>, "strings">;

export const renderQuestion = (
    question: PerseusRenderer,
    apiOptions: APIOptions = Object.freeze({}),
    extraProps?: ExtraProps,
): {
    container: HTMLElement;
    renderer: Perseus.Renderer;
    rerender: (question: PerseusRenderer, extraProps?: ExtraProps) => void;
    unmount: RenderResult["unmount"];
} => {
    setDependencies(testDependencies);
    registerAllWidgetsForTesting();

    let renderer: Perseus.Renderer | null = null;
    const {container, rerender, unmount} = render(
        <RenderStateRoot>
            <DependenciesContext.Provider value={testDependenciesV2}>
                <Renderer
                    ref={(node) => (renderer = node)}
                    question={question as any}
                    apiOptions={apiOptions}
                    extraProps={{
                        ...extraProps,
                        strings: mockStrings,
                    }}
                />
            </DependenciesContext.Provider>
        </RenderStateRoot>,
    );
    // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
    if (!renderer) {
        throw new Error(`Failed to render!`);
    }
    const renderAgain = (
        question: PerseusRenderer,
        extraProps?: ExtraProps,
    ) => {
        rerender(
            <RenderStateRoot>
                <DependenciesContext.Provider value={testDependenciesV2}>
                    <Renderer
                        ref={(node) => (renderer = node)}
                        question={question}
                        apiOptions={apiOptions}
                        extraProps={{
                            ...extraProps,
                            strings: mockStrings,
                        }}
                    />
                </DependenciesContext.Provider>
            </RenderStateRoot>,
        );
        if (!renderer) {
            throw new Error(`Failed to rerender!`);
        }
    };

    return {container, renderer, rerender: renderAgain, unmount};
};

const Renderer = React.forwardRef<
    Perseus.Renderer,
    {
        question: PerseusRenderer;
        apiOptions: APIOptions;
        extraProps?: PropsFor<typeof Perseus.Renderer>;
    }
>((props, ref) => {
    const dependencies = useDependencies();
    return (
        <Perseus.Renderer
            ref={ref}
            content={props.question.content}
            images={props.question.images}
            widgets={props.question.widgets}
            problemNum={0}
            apiOptions={props.apiOptions}
            strings={mockStrings}
            {...props.extraProps}
            {...dependencies}
        />
    );
});
