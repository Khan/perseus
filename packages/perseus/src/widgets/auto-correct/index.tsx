import React from "react";

import type {PerseusScore, WidgetExports} from "../../types";

// The props are type `any` on purpose so that this can receive props
// from any deprecated widget
type AutoCorrectProps = any;
type Rubric = any;
type UserInput = Empty;

class AutoCorrect extends React.Component<AutoCorrectProps> {
    static validate(userInput: UserInput, rubric: Rubric): PerseusScore {
        return {
            type: "points",
            earned: 0,
            total: 0,
            message: null,
        };
    }

    getUserInput: () => UserInput = () => {
        return {};
    };

    simpleValidate: (arg1: Rubric) => PerseusScore = (rubric) => {
        return AutoCorrect.validate(this.getUserInput(), rubric);
    };

    render() {
        return (
            <div
                style={{
                    backgroundColor: "gray",
                    border: "1px solid red",
                }}
            >
                <h1>Gone</h1>
                <p>The widget you are trying to reach is no longer available</p>
            </div>
        );
    }
}

export default {
    name: "auto-correct",
    displayName: "Auto Program",
    widget: AutoCorrect,
} as WidgetExports<typeof AutoCorrect>;
