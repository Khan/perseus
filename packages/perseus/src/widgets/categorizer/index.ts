export {default, Categorizer} from "./categorizer";
