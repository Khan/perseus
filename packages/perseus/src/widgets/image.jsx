/* eslint-disable react/prop-types */
/* eslint-disable react/sort-comp */
// @flow
import {StyleSheet, css} from "aphrodite";
import classNames from "classnames";
import * as React from "react";
import _ from "underscore";

import SvgImage from "../components/svg-image.jsx";
import {linterContextDefault} from "../gorgon/proptypes.js";
import * as Changeable from "../mixins/changeable.jsx";
import Renderer from "../renderer.jsx";
import {baseUnitPx} from "../styles/constants.js";
import mediaQueries from "../styles/media-queries.js";

import type {PerseusImageWidgetOptions} from "../perseus-types.js";
import type {
    ChangeFn,
    PerseusScore,
    WidgetExports,
    WidgetProps,
} from "../types.js";

const defaultBoxSize = 400;
const defaultRange = [0, 10];
const defaultBackgroundImage = {
    url: null,
    width: 0,
    height: 0,
};

const editorAlignments = ["block", "full-width"];

const DEFAULT_ALIGNMENT = "block";

type RenderProps = PerseusImageWidgetOptions; // there is no transform as part of exports
type Rubric = PerseusImageWidgetOptions;
type UserInput = null;

type ExternalProps = WidgetProps<RenderProps, Rubric>;

type Props = {|
    ...ExternalProps,
    alignment: $NonMaybeType<ExternalProps["alignment"]>,
    title: $NonMaybeType<ExternalProps["title"]>,
    range: $NonMaybeType<ExternalProps["range"]>,
    box: $NonMaybeType<ExternalProps["box"]>,
    backgroundImage: $NonMaybeType<ExternalProps["backgroundImage"]>,
    labels: $NonMaybeType<ExternalProps["labels"]>,
    alt: $NonMaybeType<ExternalProps["alt"]>,
    caption: $NonMaybeType<ExternalProps["caption"]>,
    linterContext: $NonMaybeType<ExternalProps["linterContext"]>,
|};

type DefaultProps = {|
    alignment: Props["alignment"],
    title: Props["title"],
    range: Props["range"],
    box: Props["box"],
    backgroundImage: Props["backgroundImage"],
    labels: Props["labels"],
    alt: Props["alt"],
    caption: Props["caption"],
    linterContext: Props["linterContext"],
|};

class ImageWidget extends React.Component<Props> {
    static defaultProps: DefaultProps = {
        alignment: DEFAULT_ALIGNMENT,
        title: "",
        range: [defaultRange, defaultRange],
        box: [defaultBoxSize, defaultBoxSize],
        backgroundImage: defaultBackgroundImage,
        labels: [],
        alt: "",
        caption: "",
        linterContext: linterContextDefault,
    };

    change: ChangeFn = (...args) => {
        return Changeable.change.apply(this, args);
    };

    getUserInput: () => UserInput = () => {
        return null;
    };

    simpleValidate: (Rubric) => PerseusScore = (rubric) => {
        return ImageWidget.validate(this.getUserInput(), rubric);
    };

    focus: () => void = () => {}; // no-op

    render(): React.Node {
        let image;
        let alt;
        const {apiOptions} = this.props;

        const backgroundImage = this.props.backgroundImage;

        if (backgroundImage.url) {
            image = (
                <SvgImage
                    src={backgroundImage.url}
                    alt={
                        /* alt text is formatted in a sr-only
                               div next to the image in addition to
                               the alt attribute.
                               If there is no alt text at all,
                               we don't put an alt attribute on
                               the image, so that screen readers
                               know there's something they can't
                               read there :(.
                               NOTE: React <=0.13 (maybe later)
                               has a bug where it won't ever
                               remove an attribute, so if this
                               alt node is ever defined it's
                               not removed. This is sort of
                               dangerous, but we usually re-key
                               new renderers so that they're
                               rendered from scratch anyways,
                               so this shouldn't be a problem
                               in practice right now, although
                               it will exhibit weird behaviour
                               while editing. */
                        this.props.alt
                    }
                    overrideAriaHidden={true}
                    width={backgroundImage.width}
                    height={backgroundImage.height}
                    preloader={apiOptions.imagePreloader}
                    extraGraphie={{
                        box: this.props.box,
                        range: this.props.range,
                        labels: this.props.labels,
                    }}
                    trackInteraction={this.props.trackInteraction}
                    zoomToFullSizeOnMobile={apiOptions.isMobile}
                    constrainHeight={apiOptions.isMobile}
                    allowFullBleed={apiOptions.isMobile}
                />
            );
        }

        if (this.props.alt) {
            alt = (
                <span className="perseus-sr-only">
                    <Renderer
                        content={this.props.alt}
                        apiOptions={apiOptions}
                        linterContext={this.props.linterContext}
                    />
                </span>
            );
        }

        // For mobile we combine an image's title and caption.
        if (apiOptions.isMobile) {
            let titleAndCaption;

            if (this.props.title || this.props.caption) {
                let title = this.props.title;

                // Bold the title, and make it the first sentence of the
                // caption.
                if (title) {
                    // We add a period to separate the title from the caption
                    // (if it exists), unless the title already ends with a
                    // punctuation symbol (whitespace ignored). Copied from
                    // webapp: https://github.com/Khan/webapp/blob/6e930637edb65696d0749ea0f7558214aee32b4e/javascript/tutorial-shared-package/components/content-description.jsx#L80
                    // TODO(charlie): Internationalize this check, and the
                    // delimiter that is being inserted.
                    if (this.props.caption && !/[.?!"']\s*$/.test(title)) {
                        title += ".";
                    }

                    title = `**${title}** `;
                }

                const className = classNames({
                    "perseus-image-caption": true,
                    "has-title": !!title,
                });

                // Caption is left-aligned within a container that's centered
                // below the image, with these width constraints:
                //
                // 1. Size caption to width of the image on-screen.
                // 2. ... but constrain its width to a range based on the
                //    device to optimize readability - e.g. [320px, 450px] for
                //    phones.
                // 3. ... unless the image is floated, in which case we don't
                //    want the caption to overflow the image size.
                //
                // TODO(david): If caption is only 1 line long, center-align
                //     the text.
                const alignment = this.props.alignment;
                const isImageFullWidth =
                    alignment === "block" || alignment === "full-width";

                // This minWidth takes precedence over minWidth applied via
                // Aphrodite.
                const minWidth = isImageFullWidth ? null : "0 !important";

                titleAndCaption = (
                    <div className={className}>
                        <div
                            className={css(styles.caption)}
                            style={{
                                minWidth: minWidth,
                            }}
                        >
                            <Renderer
                                content={title + this.props.caption}
                                apiOptions={apiOptions}
                                linterContext={this.props.linterContext}
                            />
                        </div>
                    </div>
                );
            }

            return (
                <div className="perseus-image-widget">
                    {image}
                    {alt}
                    {titleAndCaption}
                </div>
            );
        }
        let title;
        let caption;

        if (this.props.title) {
            title = (
                <div className="perseus-image-title">
                    <Renderer
                        content={this.props.title}
                        apiOptions={apiOptions}
                        linterContext={this.props.linterContext}
                    />
                </div>
            );
        }

        if (this.props.caption) {
            caption = (
                <div className="perseus-image-caption">
                    <Renderer
                        content={this.props.caption}
                        apiOptions={apiOptions}
                        linterContext={this.props.linterContext}
                    />
                </div>
            );
        }

        return (
            <div className="perseus-image-widget">
                {title}
                {image}
                {alt}
                {caption}
            </div>
        );
    }

    static validate(userInput: UserInput, rubric: Rubric): PerseusScore {
        return {
            type: "points",
            earned: 0,
            total: 0,
            message: null,
        };
    }
}

const styles = StyleSheet.create({
    caption: {
        display: "inline-block",
        marginTop: baseUnitPx,
        maxWidth: 640,

        [mediaQueries.lgOrSmaller]: {
            // TODO(david): This maxWidth is not being used because
            //     it's overriden by the 512px max-width we have on
            //     paragraphs.
            maxWidth: 540,
        },

        [mediaQueries.smOrSmaller]: {
            maxWidth: 450,
        },
    },
});

export default ({
    name: "image",
    // This widget's accessibility depends on its contents: if the image has
    // has a background but no alt text, it is not accessible
    accessible: (widgetOptions: PerseusImageWidgetOptions): boolean => {
        const bgImage = widgetOptions.backgroundImage;
        return !(bgImage && bgImage.url && !widgetOptions.alt);
    },
    defaultAlignment: DEFAULT_ALIGNMENT,
    supportedAlignments: editorAlignments,
    displayName: "Image",
    widget: ImageWidget,
    isLintable: true,
}: WidgetExports<typeof ImageWidget>);
