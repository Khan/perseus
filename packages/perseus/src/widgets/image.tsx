/* eslint-disable react/sort-comp */
import {linterContextDefault} from "@khanacademy/perseus-linter";
import classNames from "classnames";
import * as React from "react";
import _ from "underscore";

import AssetContext from "../asset-context";
import SvgImage from "../components/svg-image";
import * as Changeable from "../mixins/changeable";
import Renderer from "../renderer";

import type {Range, PerseusImageWidgetOptions} from "../perseus-types";
import type {
    ChangeFn,
    PerseusScore,
    WidgetExports,
    WidgetProps,
} from "../types";

const defaultBoxSize = 400;
const defaultRange: Range = [0, 10];
const defaultBackgroundImage = {
    url: null,
    width: 0,
    height: 0,
} as const;

const editorAlignments = ["block", "full-width"];

const DEFAULT_ALIGNMENT = "block";

type RenderProps = PerseusImageWidgetOptions; // there is no transform as part of exports
type Rubric = PerseusImageWidgetOptions;
type UserInput = null;

type ExternalProps = WidgetProps<RenderProps, Rubric>;

type Props = ExternalProps & {
    alignment: NonNullable<ExternalProps["alignment"]>;
    title: NonNullable<ExternalProps["title"]>;
    range: NonNullable<ExternalProps["range"]>;
    box: NonNullable<ExternalProps["box"]>;
    backgroundImage: NonNullable<ExternalProps["backgroundImage"]>;
    labels: NonNullable<ExternalProps["labels"]>;
    alt: NonNullable<ExternalProps["alt"]>;
    caption: NonNullable<ExternalProps["caption"]>;
    linterContext: NonNullable<ExternalProps["linterContext"]>;
};

type DefaultProps = {
    alignment: Props["alignment"];
    title: Props["title"];
    range: Props["range"];
    box: Props["box"];
    backgroundImage: Props["backgroundImage"];
    labels: Props["labels"];
    alt: Props["alt"];
    caption: Props["caption"];
    linterContext: Props["linterContext"];
};

class ImageWidget extends React.Component<Props> {
    static defaultProps: DefaultProps = {
        alignment: DEFAULT_ALIGNMENT,
        title: "",
        range: [defaultRange, defaultRange],
        box: [defaultBoxSize, defaultBoxSize],
        backgroundImage: defaultBackgroundImage,
        labels: [],
        alt: "",
        caption: "",
        linterContext: linterContextDefault,
    };

    change: ChangeFn = (...args) => {
        return Changeable.change.apply(this, args);
    };

    getUserInput: () => UserInput = () => {
        return null;
    };

    simpleValidate: (arg1: Rubric) => PerseusScore = (rubric) => {
        return ImageWidget.validate(this.getUserInput(), rubric);
    };

    focus: () => void = () => {}; // no-op

    render(): React.ReactNode {
        let image;
        let alt;
        const {apiOptions} = this.props;

        const backgroundImage = this.props.backgroundImage;

        if (backgroundImage.url) {
            const url = backgroundImage.url;
            image = (
                <AssetContext.Consumer>
                    {({setAssetStatus}) => (
                        <SvgImage
                            src={url}
                            alt={
                                /* alt text is formatted in a sr-only
                               div next to the image in addition to
                               the alt attribute.
                               If there is no alt text at all,
                               we don't put an alt attribute on
                               the image, so that screen readers
                               know there's something they can't
                               read there :(.
                               NOTE: React <=0.13 (maybe later)
                               has a bug where it won't ever
                               remove an attribute, so if this
                               alt node is ever defined it's
                               not removed. This is sort of
                               dangerous, but we usually re-key
                               new renderers so that they're
                               rendered from scratch anyways,
                               so this shouldn't be a problem
                               in practice right now, although
                               it will exhibit weird behaviour
                               while editing. */
                                this.props.alt
                            }
                            overrideAriaHidden={true}
                            width={backgroundImage.width}
                            height={backgroundImage.height}
                            preloader={apiOptions.imagePreloader}
                            extraGraphie={{
                                box: this.props.box,
                                range: this.props.range,
                                labels: this.props.labels,
                            }}
                            trackInteraction={this.props.trackInteraction}
                            zoomToFullSizeOnMobile={apiOptions.isMobile}
                            constrainHeight={apiOptions.isMobile}
                            allowFullBleed={apiOptions.isMobile}
                            setAssetStatus={setAssetStatus}
                        />
                    )}
                </AssetContext.Consumer>
            );
        }

        if (this.props.alt) {
            alt = (
                <span className="perseus-sr-only">
                    <Renderer
                        content={this.props.alt}
                        apiOptions={apiOptions}
                        linterContext={this.props.linterContext}
                    />
                </span>
            );
        }

        // For mobile we combine an image's title and caption.
        if (apiOptions.isMobile) {
            let titleAndCaption;

            if (this.props.title || this.props.caption) {
                let title = this.props.title;

                // Bold the title, and make it the first sentence of the
                // caption.
                if (title) {
                    // We add a period to separate the title from the caption
                    // (if it exists), unless the title already ends with a
                    // punctuation symbol (whitespace ignored). Copied from
                    // webapp: https://github.com/Khan/webapp/blob/6e930637edb65696d0749ea0f7558214aee32b4e/javascript/tutorial-shared-package/components/content-description.jsx#L80
                    // TODO(charlie): Internationalize this check, and the
                    // delimiter that is being inserted.
                    if (this.props.caption && !/[.?!"']\s*$/.test(title)) {
                        title += ".";
                    }

                    title = `**${title}** `;
                }

                const className = classNames({
                    "perseus-image-caption": true,
                    "has-title": !!title,
                });

                // Caption is left-aligned within a container that's centered
                // below the image, with these width constraints:
                //
                // 1. Size caption to width of the image on-screen.
                // 2. ... but constrain its width to a range based on the
                //    device to optimize readability - e.g. [320px, 450px] for
                //    phones.
                // 3. ... unless the image is floated, in which case we don't
                //    want the caption to overflow the image size.
                //
                // TODO(david): If caption is only 1 line long, center-align
                //     the text.
                const alignment = this.props.alignment;
                const isImageFullWidth =
                    alignment === "block" || alignment === "full-width";

                // This minWidth takes precedence over minWidth applied via
                // Aphrodite.
                const minWidth = isImageFullWidth ? null : "0 !important";

                titleAndCaption = (
                    <figcaption
                        className={className}
                        style={{
                            maxWidth: backgroundImage.width,
                        }}
                    >
                        <div
                            style={{
                                // @ts-expect-error - TS2322 - Type 'string | null' is not assignable to type 'MinWidth<string | number> | undefined'.
                                minWidth: minWidth,
                            }}
                        >
                            <Renderer
                                content={title + this.props.caption}
                                apiOptions={apiOptions}
                                linterContext={this.props.linterContext}
                            />
                        </div>
                    </figcaption>
                );
            }

            return (
                <figure
                    className="perseus-image-widget"
                    style={{
                        maxWidth: backgroundImage.width,
                    }}
                >
                    {image}
                    {alt}
                    {titleAndCaption}
                </figure>
            );
        }
        let title;
        let caption;

        if (this.props.title) {
            title = (
                <div className="perseus-image-title">
                    <Renderer
                        content={this.props.title}
                        apiOptions={apiOptions}
                        linterContext={this.props.linterContext}
                    />
                </div>
            );
        }

        if (this.props.caption) {
            caption = (
                <figcaption
                    className="perseus-image-caption"
                    style={{
                        maxWidth: backgroundImage.width,
                    }}
                >
                    <Renderer
                        content={this.props.caption}
                        apiOptions={apiOptions}
                        linterContext={this.props.linterContext}
                    />
                </figcaption>
            );
        }

        return (
            <figure
                className="perseus-image-widget"
                style={{
                    maxWidth: backgroundImage.width,
                }}
            >
                {title}
                {image}
                {alt}
                {caption}
            </figure>
        );
    }

    static validate(userInput: UserInput, rubric: Rubric): PerseusScore {
        return {
            type: "points",
            earned: 0,
            total: 0,
            message: null,
        };
    }
}

export default {
    name: "image",
    // This widget's accessibility depends on its contents: if the image has
    // has a background but no alt text, it is not accessible
    accessible: (widgetOptions: PerseusImageWidgetOptions): boolean => {
        const bgImage = widgetOptions.backgroundImage;
        return !(bgImage && bgImage.url && !widgetOptions.alt);
    },
    defaultAlignment: DEFAULT_ALIGNMENT,
    supportedAlignments: editorAlignments,
    displayName: "Image",
    widget: ImageWidget,
    isLintable: true,
} as WidgetExports<typeof ImageWidget>;
