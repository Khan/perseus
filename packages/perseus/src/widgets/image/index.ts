export {default} from "./image.class";
