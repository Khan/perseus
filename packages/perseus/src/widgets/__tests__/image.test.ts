import {describe, beforeEach, it} from "@jest/globals";

import {testDependencies} from "../../../../../testing/test-dependencies";
import * as Dependencies from "../../dependencies";
import {isAccessible} from "../../widgets";
import {question} from "../__testdata__/image.testdata";

import {renderQuestion} from "./renderQuestion";

import type {ImageWidget, PerseusImageWidgetOptions} from "../../perseus-types";
import type {APIOptions} from "../../types";

describe.each([true, false])("image widget - isMobile %b", (isMobile) => {
    const apiOptions: APIOptions = {isMobile};

    beforeEach(() => {
        jest.spyOn(Dependencies, "getDependencies").mockReturnValue(
            testDependencies,
        );
    });

    it("should snapshot", () => {
        // Arrange

        // Act
        const {container} = renderQuestion(question, apiOptions);

        // Assert
        expect(container).toMatchSnapshot("first render");
    });

    it("should be unanswerable", () => {
        // Arrange

        // Act
        const {renderer} = renderQuestion(question, apiOptions);

        // Assert
        expect(renderer).toHaveBeenAnsweredIncorrectly();
    });

    it("should be accessible if background has 'alt' prop", () => {
        // Arrange, Act, and Assert
        expect(isAccessible(question.widgets["image 1"])).toBeTrue();
    });

    it("should be inaccessible if background is missing 'alt' prop", () => {
        // Arrange
        const imageWidget = question.widgets["image 1"] as ImageWidget;
        const options = imageWidget.options as PerseusImageWidgetOptions;
        const inaccessibleWidgetInfo = {
            ...imageWidget,
            options: {
                ...options,
                alt: "",
            },
        } as ImageWidget;

        // Act and Assert
        expect(isAccessible(inaccessibleWidgetInfo)).toBeFalse();
    });
});
