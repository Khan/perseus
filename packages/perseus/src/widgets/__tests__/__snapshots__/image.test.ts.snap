// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`image widget - isMobile %b should snapshot: first render 1`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive perseus-renderer-two-columns"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="perseus-two-columns"
      >
        <div
          class="perseus-column"
        >
          <div
            class="perseus-column-content"
          >
            <div
              class="paragraph"
            >
              <div
                class="perseus-widget-container widget-nohighlight widget-block"
              >
                <figure
                  class="perseus-image-widget"
                  style="max-width: 420px;"
                >
                  <div
                    class="fixed-to-responsive svg-image"
                    style="max-width: 420px; max-height: 345px;"
                  >
                    <div
                      style="padding-bottom: 82.14%;"
                    />
                    <span
                      style="top: 0px; left: 0px; width: 100%; height: 100%; position: absolute; min-width: 20px; display: flex; justify-content: center; align-content: center;"
                    >
                      <div
                        class="default_xu2jcg-o_O-spinnerContainer_agrn11"
                      >
                        <svg
                          height="48"
                          viewBox="0 0 48 48"
                          width="48"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            class="loadingSpinner_1dvgeb3-o_O-inlineStyles_1kzk0y9"
                            d="M44.19 23.455a1.91 1.91 0 1 1 3.801 0h.003c.004.18.006.363.006.545 0 13.255-10.745 24-24 24S0 37.255 0 24 10.745 0 24 0c.182 0 .364.002.545.006V.01a1.91 1.91 0 1 1 0 3.801v.015A20.564 20.564 0 0 0 24 3.818C12.854 3.818 3.818 12.854 3.818 24c0 11.146 9.036 20.182 20.182 20.182 11.146 0 20.182-9.036 20.182-20.182 0-.182-.003-.364-.007-.545h.015z"
                            fill-rule="nonzero"
                          />
                        </svg>
                      </div>
                    </span>
                  </div>
                  <span
                    class="perseus-sr-only"
                  >
                    <div
                      class="perseus-renderer perseus-renderer-responsive"
                    >
                      <div
                        class="paragraph"
                        data-perseus-paragraph-index="0"
                      >
                        <div
                          class="paragraph"
                        >
                          An array of isosceles triangles. A triangle has height A. Two smaller triangle, one with height B and one with height C, have approximately the same combined height as A.
                        </div>
                      </div>
                    </div>
                  </span>
                  <figcaption
                    class="perseus-image-caption has-title"
                    style="max-width: 420px;"
                  >
                    <div>
                      <div
                        class="perseus-renderer perseus-renderer-responsive"
                      >
                        <div
                          class="paragraph"
                          data-perseus-paragraph-index="0"
                        >
                          <div
                            class="paragraph"
                          >
                            <strong>
                              Image Title.
                            </strong>
                             Image Caption
                          </div>
                        </div>
                      </div>
                    </div>
                  </figcaption>
                </figure>
              </div>
            </div>
          </div>
        </div>
        <div
          class="perseus-column"
        >
          <div
            class="perseus-column-content"
          >
            <div
              class="paragraph"
            >
              A quilter wants to make the design shown at left using the Golden Ratio. Specifically, he wants the ratio of the triangle heights 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  A:B
                </span>
                <span />
              </span>
               and 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  B:C
                </span>
                <span />
              </span>
               to each equal 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  1.62
                </span>
                <span />
              </span>
              . If the quilter makes the triangle height 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  A=8\\ \\text{in}
                </span>
                <span />
              </span>
              , approximately how tall should he make triangle height 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  C
                </span>
                <span />
              </span>
              ?
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`image widget - isMobile %b should snapshot: first render 2`] = `
<div>
  <div
    class="perseus-renderer perseus-renderer-responsive perseus-renderer-two-columns"
  >
    <div
      class="paragraph"
      data-perseus-paragraph-index="0"
    >
      <div
        class="perseus-two-columns"
      >
        <div
          class="perseus-column"
        >
          <div
            class="perseus-column-content"
          >
            <div
              class="paragraph"
            >
              <div
                class="perseus-widget-container widget-nohighlight widget-block"
              >
                <figure
                  class="perseus-image-widget"
                  style="max-width: 420px;"
                >
                  <div
                    class="perseus-image-title"
                  >
                    <div
                      class="perseus-renderer perseus-renderer-responsive"
                    >
                      <div
                        class="paragraph"
                        data-perseus-paragraph-index="0"
                      >
                        <div
                          class="paragraph"
                        >
                          Image Title
                        </div>
                      </div>
                    </div>
                  </div>
                  <div
                    class="fixed-to-responsive svg-image"
                    style="max-width: 420px; max-height: 345px;"
                  >
                    <div
                      style="padding-bottom: 82.14%;"
                    />
                    <span
                      style="top: 0px; left: 0px; width: 100%; height: 100%; position: absolute; min-width: 20px; display: flex; justify-content: center; align-content: center;"
                    >
                      <div
                        class="default_xu2jcg-o_O-spinnerContainer_agrn11"
                      >
                        <svg
                          height="48"
                          viewBox="0 0 48 48"
                          width="48"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            class="loadingSpinner_1dvgeb3-o_O-inlineStyles_1kzk0y9"
                            d="M44.19 23.455a1.91 1.91 0 1 1 3.801 0h.003c.004.18.006.363.006.545 0 13.255-10.745 24-24 24S0 37.255 0 24 10.745 0 24 0c.182 0 .364.002.545.006V.01a1.91 1.91 0 1 1 0 3.801v.015A20.564 20.564 0 0 0 24 3.818C12.854 3.818 3.818 12.854 3.818 24c0 11.146 9.036 20.182 20.182 20.182 11.146 0 20.182-9.036 20.182-20.182 0-.182-.003-.364-.007-.545h.015z"
                            fill-rule="nonzero"
                          />
                        </svg>
                      </div>
                    </span>
                  </div>
                  <span
                    class="perseus-sr-only"
                  >
                    <div
                      class="perseus-renderer perseus-renderer-responsive"
                    >
                      <div
                        class="paragraph"
                        data-perseus-paragraph-index="0"
                      >
                        <div
                          class="paragraph"
                        >
                          An array of isosceles triangles. A triangle has height A. Two smaller triangle, one with height B and one with height C, have approximately the same combined height as A.
                        </div>
                      </div>
                    </div>
                  </span>
                  <figcaption
                    class="perseus-image-caption"
                    style="max-width: 420px;"
                  >
                    <div
                      class="perseus-renderer perseus-renderer-responsive"
                    >
                      <div
                        class="paragraph"
                        data-perseus-paragraph-index="0"
                      >
                        <div
                          class="paragraph"
                        >
                          Image Caption
                        </div>
                      </div>
                    </div>
                  </figcaption>
                </figure>
              </div>
            </div>
          </div>
        </div>
        <div
          class="perseus-column"
        >
          <div
            class="perseus-column-content"
          >
            <div
              class="paragraph"
            >
              A quilter wants to make the design shown at left using the Golden Ratio. Specifically, he wants the ratio of the triangle heights 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  A:B
                </span>
                <span />
              </span>
               and 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  B:C
                </span>
                <span />
              </span>
               to each equal 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  1.62
                </span>
                <span />
              </span>
              . If the quilter makes the triangle height 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  A=8\\ \\text{in}
                </span>
                <span />
              </span>
              , approximately how tall should he make triangle height 
              <span
                style="white-space: nowrap;"
              >
                <span />
                <span
                  class="mock-TeX"
                >
                  C
                </span>
                <span />
              </span>
              ?
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
